### ISSUE TITLE
Parse.Cloud.beforeLogin does not fire when a user is signed up and logged in for the first time

### ISSUE THREAD

**Issue by estradanic** [parse-community/parse-server#8211]
### New Issue Checklist

- [x] I am not disclosing a [vulnerability](https://github.com/parse-community/parse-server/blob/master/SECURITY.md).
- [x] I am not just asking a [question](https://github.com/parse-community/.github/blob/master/SUPPORT.md).
- [x] I have searched through [existing issues](https://github.com/parse-community/parse-server/issues?q=is%3Aissue).
- [ ] I can reproduce the issue with the [latest version of Parse Server](https://github.com/parse-community/parse-server/releases).

### Issue Description
`beforeLogin` trigger does not fire on signup.

### Steps to reproduce
Create a `beforeLogin` trigger, and inside of it, a `console.log`, then use Parse.User.signUp() to create a new user (thus automatically logging them in)

### Actual Outcome
The trigger is not called; the log is never printed to the console.

### Expected Outcome
The trigger is called, because signUp not only creates the user. It also creates a session and logs them in.

### Environment
4.5.0

Server
- Parse Server version: 4.5.0
- Operating system: Unknown
- Local or remote host (AWS, Azure, Google Cloud, Heroku, Digital Ocean, etc): Back4app
- Database System (MongoDB or Postgres): MongoDB
- Database version: 3.6
- Local or remote host (MongoDB Atlas, mLab, AWS, Azure, Google Cloud, etc): Back4app

Client
- SDK (iOS, Android, JavaScript, PHP, Unity, etc): JavaScript
- SDK version: 3.4.4

**Comment by mtrezza** [comment]
> The trigger is called, because signUp not only creates the user. It also creates a session and logs them in.

That is a good point. Could you please make sure this issue still exists in the latest Parse Server 5.x-alpha version?

**Comment by estradanic** [comment]
I'll try, but it might take me some time. I'm using parse for a side project that I have on back4app.com, and they only have versions up to 4.5.0 available for use as of now. I'll have to squeeze in a local install some time after work.

**Comment by mtrezza** [comment]
If we can get a PR merged by this month it would be easier to write the PR. After that we'd have to add a deprecation mechanism because this is a breaking change.

**Comment by estradanic** [comment]
I've had a real hard time getting parse up and running locally in the bare minutes I have between work and home-life. Back4App is now beta-testing Parse 5 for their platform, so hopefully I get a response from them soon and I can test it there to make sure it's still broken on the latest version.

**Comment by mtrezza** [comment]
It seems what you are experiencing is expected behavior. There is a test case for it:

https://github.com/parse-community/parse-server/blob/81304be649306708fe63a26ae828edcdf4d0ad17/spec/CloudCode.spec.js#L3096-L3107

If you can make a case why this should be triggered also on sign-up, then we may be able to change the behavior and release it with Parse Server 6, for which we are currently preparing the alpha version. For Parse Server 5 this is unlikely to change as it's already moving to long-term support in 2 months.

**Comment by estradanic** [comment]
The reason I think it should be triggered with signUp is because the primary use case for beforeLogin is to ensure some code runs before session creation. The primary example I've seen around the web is to for example, check a banned list of users to prevent login. 

signUp also creates a session and logs the user in, so beforeLogin should run. I want to create a whitelist of users who can sign up and the only way to do that currently is either solely client-side limited or just janky (got to intercept the signUp request with node, check it, and only allow it through if the whitelist entry is present)

**Comment by dblythy** [comment]
I think the current solution is to use a `beforeSave` trigger on `Parse.User` and run the code on `req.object.existed()`

We could possibly introduce a `beforeCreate` trigger that could allow an easier way to handle and reject new objects/users

**Comment by dblythy** [comment]
A solution could be to allow a more `CRUD` style cloud hooks

- add a beforeCreate/afterCreate trigger that fires on new objects/files/users (only fire with existing objects)
- change the beforeSave/afterSave triggers to beforeUpdate/afterUpdate
- remove/depreciate beforeLogin/afterLogin and replace with beforeCreate(Parse.Session) / afterCreate(Parse.Session)

**Comment by estradanic** [comment]
That's a possibility, but I'm not sure it lines up well with the SDK's functions. 

.save() -> .beforeSave()

.logIn() -> .beforeLogin()

There is no .create() or .update(). 

I think changing the trigger names sparks a lot of further conversation. A solution more consistent with the SDKs would be to add a beforeSignUp trigger. 

I really like the idea of parse as a whole becoming more transparent with its operations, but it seems a bigger conversation than this item.

**Comment by mtrezza** [comment]
Would a beforeSave trigger on `_Session` be what you're looking for?

**Comment by estradanic** [comment]
I'll try it out and see if that works! I have a feeling it will, but I didn't think of it!

**Comment by estradanic** [comment]
No dice there. The cloud code failed to load in the first place until I removed the beforeSave. 

Got it working by using a Parse.User beforeSave and checking isNew on the request object

**Comment by mtrezza** [comment]
> No dice there. The cloud code failed to load in the first place until I removed the beforeSave.

Not sure but trigger on _Session may not be allowed currently. At least it wasn't some time ago.

> Got it working by using a Parse.User beforeSave and checking isNew on the request object

Is that triggered on login?

All in all I think the triggers could be made easier to grasp. A trigger diagram would certainly be helpful, so see where which trigger is invoked in the process.

**Comment by estradanic** [comment]
> Is that triggered on login?

No, not on login. I'm using `beforeLogin()` and `beforeSave(Parse.User)` together to cover all cases. Not great, but functional.

> All in all I think the triggers could be made easier to grasp. A trigger diagram would certainly be helpful, so see where which trigger is invoked in the process.

Yes, I think that would be helpful.
